agent/command.go

```diff
@@ -165,12 +165,20 @@ func (a *Agent) runCommand(ctx context.Context, tc *taskContext, logger client.L
 	}
 	defer func() {
 		if !tc.unsetFunctionVarsDisabled || tc.project.UnsetFunctionVars {
-			// This defer ensures that the function vars do not persist in the expansions after the function is over.
-			for key, functionValue := range commandInfo.Vars {
-				currentValue := tc.taskConfig.Expansions.Get(key)
-				if currentValue != functionValue {
-					// If a command in the func updates the expansion value, don't reset it.
-					delete(prevExp, key)
+			// This defer ensures that the function vars do not persist in the expansions after the function is over
+			// unless they were updated using expansions.update
+
+			if cmd.Name() == "expansions.update" {
+				c, ok := cmd.(*command.Update)
```

Comment:
I don't think this will work if the expansions are specified via a YAML file rather than explicitly listing them in `updates`. This is also a little janky because we're closely depending on the exact way that `expansions.update` is implemented and trying to reproduce its behavior here again.

Instead of doing it like this, can we consider an alternative implementation? One idea is adding a `TaskConfig.DynamicExpansions` field and modifying `expansions.update` to duplicate its writes there (and then unsetting the dynamic expansions when `runCommand` returns). Then, so long as `expansions.update` replicates its updates to the dynamic expansions, it gives a way for this defer to read the updated expansions, without relying on how exactly `expansions.update` is implemented or awkwardly type-casting.